geopandas/tools/overlay.py

```diff
@@ -176,3 +177,177 @@ def overlay(df1, df2, how, use_sindex=True):
 
     # Return geodataframe with new indices
     return GeoDataFrame(collection, index=range(len(collection)))
+
+
+def overlay_intersection(df1, df2):
+    """
+    Overlay Intersection operation used in overlay function
+    """
+    # Spatial Index to create intersections
+    spatial_index = df2.sindex
+    bbox = df1.geometry.apply(lambda x: x.bounds)
+    sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
+    # Create pairs of geometries in both dataframes to be intersected
+    nei = []
+    for i, j in sidx.items():
+        for k in j:
+            nei.append([i, k])
+    if nei != []:
+        pairs = pd.DataFrame(nei, columns=['idx1','idx2'])
+        left = df1.geometry.take(pairs['idx1'].values)
+        left.reset_index(drop=True, inplace=True)
+        right = df2.geometry.take(pairs['idx2'].values)
+        right.reset_index(drop=True, inplace=True)
+        intersections = left.intersection(right).buffer(0)
+
+        # only keep actual intersecting geometries
+        pairs_intersect = pairs[~intersections.is_empty]
+        geom_intersect = intersections[~intersections.is_empty]
```

Comment:
@ozak I kept the logic, but changed the implementation a little bit. So instead of first merging all data, calculating the new geoms and then removing rows, I now first calculate the geoms, and then only merge those parts of the data that we need (this makes it a very little bit more efficient). 
I also replaced the apply intersection with an `intersection` method call, so this will be speed up further with the cythonization.